lib/spack/spack/cmd/__init__.py

```diff
@@ -177,6 +177,30 @@ def parse_specs(
     if not concretize:
         return specs
 
+    # Special case for concretizing a single spec
+    if len(specs) == 1:
+        return specs[0].concretized()
+
+    # Special case if every spec has an abstract spec
+    if all(spec.concrete or spec.abstract_hash for spec in specs):
+        ret = [s.concretize() for s in specs]
+
+        unify = spack.config.get("concretizer:unify", False)
+        if unify is True:
+            hashes_per_name = {}
+            for spec in ret:
+                for dep in spec.traverse(deptype=("link", "run")):
+                    if dep.name == "gcc-runtime":
+                        # gcc runtime is allowed multiple times in a DAG unlike other pkgs
+                        continue
+                    specs_per_name.set_default(dep.name, set()).append(dep.dag_hash())
+            if any(len(hashes) > 1 for _, hashes in hashes_per_name.items()):
```

Comment:
I've found this thing to be useful and very clear:

```suggestion
            from collections import Counter
            
            specs_per_name = Counter(
                rep.name
                for spec in ret for dep in spec.traverse(deptype=("link", "run"))
                if dep.name != "gcc-runtime"  # gcc allowed multiple times in runtime DAG
            )
            
            if any(count > 1 for count in specs_per_name.values()):
```